src/board.rs

```diff
@@ -0,0 +1,1021 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+use crate::judge::Judge;
+
+#[derive(EnumIter, Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Direction {
+    South,
+    East,
+    North,
+    West,
+}
+
+impl Direction {
+    fn add(self, point: Coordinate) -> Coordinate {
+        match self {
+            Direction::North => Coordinate {
+                x: point.x,
+                y: point.y + -1, // We use the computer graphics convention of (0,0) in the top left
+            },
+            Direction::South => Coordinate {
+                x: point.x,
+                y: point.y + 1,
+            },
+            Direction::East => Coordinate {
+                x: point.x + 1,
+                y: point.y,
+            },
+            Direction::West => Coordinate {
+                x: point.x + -1,
+                y: point.y,
+            },
+        }
+    }
+
+    // Returns whether vertical words should be read from top to bottom if played by a player on this side of the board
+    fn read_top_to_bottom(self) -> bool {
+        matches!(self, Direction::South) || matches!(self, Direction::West)
+    }
+
+    // Returns whether horizontal words should be read from left to right if played by a player on this side of the board
+    fn read_left_to_right(self) -> bool {
+        matches!(self, Direction::South) || matches!(self, Direction::East)
+    }
+
+    pub fn opposite(self) -> Self {
+        match self {
+            Direction::North => Direction::South,
+            Direction::South => Direction::North,
+            Direction::East => Direction::West,
+            Direction::West => Direction::East,
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Board {
+    squares: Vec<Vec<Option<Square>>>,
+    roots: Vec<Coordinate>,
+    orientations: Vec<Direction>, // The side of the board that the player is sitting at, and the direction that their vertical words go in
+}
+
+impl Board {
+    pub fn new(width: usize, height: usize) -> Self {
+        // TODO: is all this internal usize <-> isize conversion worth accepting isize as valid coordinates? Is that only used for simpler traversal algorithms?
```

Comment:
Makes methods like finding the neighbours of a square more convenient. You just look at the coordinates on each of the 4 sides of the square without taking edge/corner cases into account, because they're handled by lower level methods.

Reply:
imo I'd put that abstraction somewhere else under the "make invalid states un-representable" motto of Rust

Reply:
But actually, I think I can rework it to remove negatives and keep the same behaviour. I'll just allow wrapping in `coordinate.add(direction)`, and the `board.get(coordinate)` will still fail for "negative" coordinates, since it'll wrap to a high invalid number.